addons/website/controllers/main.py

```diff
@@ -849,6 +877,55 @@ def theme_customize_bundle_reload(self):
     # Server actions
     # ------------------------------------------------------
 
+    @http.route(['/website/theme_upload_font'], type='json', auth='user', website=True)
+    def theme_upload_font(self, name, mimetype, data):
+        """
+        Uploads font binary data and returns metadata about accessing individual fonts.
+        :param name: name of the uploaded file
+        :param mimetype: mimetype of the uploaded file
+        :param data: binary content of the uploaded file
+        :return: list of dict describing each contained font with:
+            - name
+            - variant_name
+            - attachment URL
+            - mimetype
+        """
+        def create_attachment(font, data):
+            """ Creates font attachments right away to avoid keeping
+            several extracted contents in memory. """
+            ext = font['name'].split('.')[-1].lower()
+            font['mimetype'] = f'font/{ext}'
+            attachment = request.env['ir.attachment'].create({
+                'name': font['name'],
+                'mimetype': font['mimetype'],
+                'datas': data,
+                'public': True,
+            })
+            font['id'] = attachment.id
+            font['url'] = f"/web/content/{attachment.id}/{font['name']}"
+            return font
+
+        result = []
+        binary_data = base64.b64decode(data)
+        readable_data = BytesIO(binary_data)
+        if zipfile.is_zipfile(readable_data):
+            with zipfile.ZipFile(readable_data, "r") as zip_file:
+                for entry in zip_file.filelist:
+                    if entry.file_size > MAX_FONT_FILE_SIZE:
+                        raise UserError(_("File '%s' exceeds maximum allowed file size", entry.filename))
+                    if entry.filename.split('.')[-1].lower() not in SUPPORTED_FONT_EXTENSIONS:
+                        continue
+                    result.append(create_attachment({
+                        'name': f'{name}-{entry.filename}',
+                    }, base64.b64encode(zip_file.read(entry))))
+        elif name.split('.')[-1].lower() in SUPPORTED_FONT_EXTENSIONS:
```

Comment:
That means you can upload whatever you want as long as the extension looks like a font, and the uploader never complains (just tried with a pdf renamed .ttf). Shouldn't you check the true mimetype at least?

Reply:
Added a second commit with a check based on the file content.
Did not do it inside the general `guess_mimetype` because that one potentially returns result from a library instead of our own implementation - and those results do not match what we expect here.